Autonomous smart contracts (#99)
massalabs/massa-docs: web3-dev/smart-contracts/autonomous-smart-contract.rst

multiple messages created in the same slot.

All the information are given by the smart-contract calling the ABI
in ABI parameters. Let's investigate the parameters in the prototype of the ABI 
`sendMessage(at, functionName, validityStartPeriod, validityStartThread, validityEndPeriod, validityEndThread, maxGas, rawFee, coins, msg, filterAddress, filterKey)` :

- `at`: Address of the smart-contract that you want to wakeup(can be yourself)
- `functionName`: Function of the smart-contract to call
- `validityStartPeriod` and `validityStartThread`: First slot where your message could be executed
- `validityEndPeriod` and `validityEndThread`: Last slot where you message can be executed
- `maxGas`: Maximum amount of gas for the execution of the function referenced in the message
- `rawFee`: A fee to gain prority in the async pool
- `coins`: Coins that will be passed to the smart contract called in the message
- `data`: Parameter of the function called in the message
- `filterAddress`: An optional address that you can define if you want your message to be executed in your validity period
but only after a modification in the data of the address (balance, bytecode, datastore)

> - `filterKey`: An optional datastore key that will execute the message only if there is changes on this datastore key of the `filterAddress`

Reviewer: Viewing this on github the rendering is confusing. This all appears on one line, as a standalone paragraph.

If this is intended to be viewed on github, it will need a fix. If the intended platform to be viewed renders appropriately, feel free to resolve yourself @AurelienFT

> - `filterKey`: An optional datastore key that will execute the message only if there is changes on this datastore key of the `filterAddress`

Author: You can see the render for this PR here : https://massa--99.org.readthedocs.build/en/99/

> - `filterKey`: An optional datastore key that will execute the message only if there is changes on this datastore key of the `filterAddress`

Reviewer: ![image](https://user-images.githubusercontent.com/16680090/218489112-bff42ad7-7076-450e-ae42-41d740ae9e3e.png)

Yeah, the rendering seems a bit buggy. I assume that `filterKey` is supposed to be a final dotpoint, with the `BUT only after modification [...]` is intended to provide additional context to `filterAddress`?

> - `filterKey`: An optional datastore key that will execute the message only if there is changes on this datastore key of the `filterAddress`

Author: Ah your right. Will modify it.

.. _autonomous-smart-contract:

Autonomous Smart Contract
=========================

Massa smart contracts have the unique feature of being able to autonomously define when they will be active, without external actors. By the end of this page, you will have a basic understanding of

- The mechanisms behind autonomous smart contracts
- How to use the sendMessage(...) ABI in order to make your smart contracts autonomous

> - Provided autonomous smart contracts examples

Author: I think it's better to have it at the end and with a proper section to add more in the future

Mechanism in Massa network
--------------------------

> This system works by emitting messages which will schedule an execution. To do so, the address of the SC to be executed, the function to executed, the desired execution window, and potential execution triggers must be specified.

Reviewer: ```suggestion
This system works by emitting messages which will schedule an execution. To do so, the following must be specified:

- The address of the SC to be executed
- The function to be executed
- The desired execution window
- The activation triggers
```

Maybe have the last point be: "- (Optional) The activation triggers"?

With autonomous SCs, the message is emitted inside of the operation and and you pay the gas for the message execution upfront. The message is then sent to another pool called the async pool.

This pool is deterministic by nature, as it's filled with information that comes from operations included in blocks which is processed by every node of the network.

The pool has a finite size, and sorts its messages with a similar behaviour as the operation pool: there is a priority, which is computed from the message emission fees, maximum execution gas, and length of the execution window. The messages get removed after a certain number of slots if they were not executed.

The number of messages that can be executed by each slot are limited by a maximum amount of gas like normal operations 
and so it's possible that your message isn't executed at the slot that you want (such as operations that could be not included in the first block after you sent it),
but in a later slot, when there is enough space to include your message. If you want to have your message to be included as soon as possible, 
you have to put more fees than the others (just like the operations).

> The order of the messages is determined by the formula: `(Reverse(Ratio(msg.fee, max(msg.max_gas,1))), emission_slot, emission_index)` where `emission_index` is an index that differentiate

Reviewer 2: I'd put this formula on a new line. Otherwise it can be cut in the middle and it's not easy to read.

multiple messages created in the same slot.

Use sendMessage to add autonomy in your smart contracts
-------------------------------------------------------

> You need to use the ABI `sendMessage(...)` in your smart contract if you want to emit a message to call a function in the future.
> 
> All information is provided by the smart-contract calling the ABI
> in ABI parameters. Let's investigate the parameters in the prototype of the ABI 
> `sendMessage(at, functionName, validityStartPeriod, validityStartThread, validityEndPeriod, validityEndThread, maxGas, rawFee, coins, msg, filterAddress, filterKey)` :

Reviewer: ```suggestion

As we've just seen, the core principal enabling the autonomous smart contracts feature, is the ability to emit a special message which has the effect of executing a desired function call at a specified window in time, and (optionally) when other specified conditions are met. 

The ABI method `sendMessage(...)` is the key to emitting such a message, and the provided parameters define its configuration: 
```

I think having the full content of line 39 is redundant to the dot-point list, especially when the first is _immediately_ followed by the second.

Reviewer 2: Agree.

- `at`: Address of the smart-contract that you want to wakeup(can be yourself)
- `functionName`: Function of the smart-contract to call
- `validityStartPeriod` and `validityStartThread`: First slot where your message could be executed
- `validityEndPeriod` and `validityEndThread`: Last slot where you message can be executed
- `maxGas`: Maximum amount of gas for the execution of the function referenced in the message
- `rawFee`: A fee to gain prority in the async pool
- `coins`: Coins that will be passed to the smart contract called in the message
- `data`: Parameter of the function called in the message
- `filterAddress`: An optional address that you can define if you want your function to be executed in your validity period

> BUT only after a modification in the data of the address (balance, bytecode, datastore)

Reviewer 2: There's a rendering problem. You should add space after the newline.

- `filterKey`: An optional datastore key that will execute the message only if there is changes on this datastore key of the `filterAddress`

Examples
--------

> You can find an example of autonomous smart contracts in `this oracle example <https://github.com/massalabs/massa-sc-examples/tree/oracle_example/oracle>`__

Reviewer 2: It's not called oracle anymore (and it has been merged).

.. _autonomous-smart-contract:

Autonomous Smart Contract
=========================

> Massa smart contracts have a unique feature of being able to autonomously define when they will be active, without external direction. By the end of this page, you will have a basic understanding of

Reviewer 3: ```suggestion
Massa smart contracts have the unique feature of being able to autonomously define when they will be active, without external actors. By the end of this page, you will have a basic understanding of
```

> - Understand the mechanism behind autonomous smart contracts
> - How to the sendMessage(...) ABI to add autonomy to your smart contracts.
> - Examples of autonomous smart contracts

Reviewer 3: ```suggestion
- The mechanisms behind autonomous smart contracts
- How to use the sendMessage(...) ABI in order to make your smart contracts autonomous
- Provided autonomous smart contracts examples
```

Mechanism in Massa network
--------------------------

> This system works by emitting messages specifying the address and the function and the you want to trigger at a certain time.

Reviewer 3: ```suggestion
This system works by emitting messages which will schedule an execution. To do so, the address of the SC to be executed, the function to executed, the desired execution window, and potential execution triggers must be specified.
```

Normally, when you submit an operation, the operation is sent in the operation pool and in the future included in a block to be executed: you pay the gas at execution time.

> With autonomous SCs, the message is emitted inside an operation and so you pay the gas for the message directly in the operation and the message is sent to another pool called async pool.

Reviewer 3: ```suggestion
With autonomous SCs, the message is emitted inside of the operation and and you pay the gas for the message execution upfront. The message is then sent to another pool called the async pool.
```

> This pool is deterministic in nature, as it's filled by information that come from operations included in blocks
> and processed by the whole network.

Reviewer 3: ```suggestion
This pool is deterministic by nature, as it's filled with information that comes from operations included in blocks which is processed by every node of the network.
```

> The pool has a finite size, and sorts its messages with the same behavior as the operation pool: there is a priority with the fees
> and the messages get removed after a certain number of slots. 

Reviewer 3: ```suggestion
The pool has a finite size, and sorts its messages with a similar behaviour as the operation pool: there is a priority, which is computed from the message emission fees, maximum execution gas, and length of the execution window. The messages get removed after a certain number of slots if they were not executed.
```

.. _autonomous-smart-contract:

Autonomous Smart Contract
=========================

Massa smart contracts have the unique feature of being able to autonomously define when they will be active, without external actors. By the end of this page, you will have a basic understanding of

- The mechanisms behind autonomous smart contracts
- How to use the sendMessage(...) ABI in order to make your smart contracts autonomous
- Provided autonomous smart contracts examples

Mechanism in Massa network
--------------------------

This system works by emitting messages which will schedule an execution. To do so, the following must be specified:

- The address of the SC to be executed
- The function to be executed
- The desired execution window
- (Optional) The activation triggers

Normally, when you submit an operation, the operation is sent in the operation pool and in the future included in a block to be executed: you pay the gas at execution time.
With autonomous SCs, the message is emitted inside of the operation and and you pay the gas for the message execution upfront. The message is then sent to another pool called the async pool.

> This pool is deterministic by nature, as it's filled with information that comes from operations included in blocks which is processed by every node of the network.

Reviewer: ```suggestion
This pool is deterministic in nature, as it's filled with information that comes from operations included in blocks which is processed by every node of the network.
```

both are correct, but to me, this change reads more naturally.

Alternatively: "The content of this pool is inherently deterministic, as it's..."

...but this suggestion is optional to the PR. Feel free to close without change.

The ABI method `sendMessage(...)` is the key to emitting such a message, and the provided parameters define its configuration: 

- `at`: Address of the smart-contract that you want to wakeup(can be yourself)
- `functionName`: Function of the smart-contract to call
- `validityStartPeriod` and `validityStartThread`: First slot where your message could be executed
- `validityEndPeriod` and `validityEndThread`: Last slot where you message can be executed
- `maxGas`: Maximum amount of gas for the execution of the function referenced in the message
- `rawFee`: A fee to gain prority in the async pool
- `coins`: Coins that will be passed to the smart contract called in the message
- `data`: Parameter of the function called in the message
- `filterAddress`: An optional address that you can define if you want your function to be executed in your validity period BUT only after a modification in the data of the address (balance, bytecode, datastore)
- `filterKey`: An optional datastore key that will execute the message only if there is changes on this datastore key of the `filterAddress`

Examples
--------

> You can find an example of autonomous smart contracts in `this ticker example <https://github.com/massalabs/massa-sc-examples/tree/main/ticker>`__

Reviewer: ```suggestion
You can find an example of autonomous smart contracts in `this ticker example <https://github.com/massalabs/massa-sc-examples/tree/main/ticker>`__, where `sendMessage(...)` is used `here <https://github.com/massalabs/massa-sc-examples/blob/a69dd88079caf0932e638689a518627b15a4d407/ticker/assembly/contracts/oracle.ts#L25>`__
```

This addition might be useful. First time exploring the link, it took me a few moments to work out where it might be. Might be useful for non-devs that are at least interested in exploring the example.

Autonomous Smart Contract
=========================

> In smart-contract world of Massa, you can create smart-contracts that can be triggered in the future.

Reviewer 2: ```suggestion
In Massa, you can create smart-contracts that can be triggered automatically in the future.
```

Reviewer: "Resolved" from my point of view

> This system works by sending messages in the future specifying the address and the function you want to trigger.

Reviewer 2: You don't really "send" the message in the future.

Reviewer 3: ```suggestion
This system allows an SC to schedule future executions. Address, executed functions and execution time range must be specified.
```

> You need to use the ABI `sendMessage(...)` in your smart contract if you want to send a message to wakeup a function in the future.

Reviewer 2: I think this can be improved as well.

Reviewer 3: I think we should use "emit async message = schedule a future execution" as a general vocabulary rule, instead of "waking up" and "sending".

When the message is sent it will debit you the gas directly in this operation and store this message in a separate pool
than the operation one called async pool.

> This pool is totally deterministic because it's filled by information that come from operations included in blocks

Reviewer 2: I think you are mixing general information (this sentence) with specific ones (ABI sendMessage above). I would restructure the file to separate things better.

and processed by the whole network.

> The pool has a finite size and sort the messages with the same behavior as the operation pool: there is a priority with the fees

Reviewer 2: Same as above.

`sendMessage(at, functionName, validityStartPeriod, validityStartThread, validityEndPeriod, validityEndThread, maxGas, rawFee, coins, msg, filterAddress, filterKey)` :

- `at`: Address of the smart-contract that you want to wakeup(can be yourself)
- `functionName`: Function of the smart-contract to call
- `validityStartPeriod` and `validityStartThread`: First slot where your message could be executed
- `validityEndPeriod` and `validityEndThread`: Last slot where you message can be executed
- `maxGas`: Maximum amount of gas for the execution of the function referenced in the message
- `rawFee`: A fee to gain prority in the async pool
- `coins`: Coins that will be passed to the smart contract called in the message
- `data`: Parameter of the function called in the message
- `filterAddress`: An optional address that you can define if you want your message to be executed in your validity period
but only after a modification in the data of the address (balance, bytecode, datastore)
- `filterKey`: An optional datastore key that will execute the message only if there is changes on this datastore key of the `filterAddress`

> You can find an example of an oracle using autonomous smart contracts `here <https://github.com/massalabs/massa-sc-examples/tree/oracle_example/oracle>`__

Reviewer 2: ```suggestion
You can find an example of autonomous smart contracts in `this oracle example <https://github.com/massalabs/massa-sc-examples/tree/oracle_example/oracle>`__
```
